titus-common-client/src/main/java/com/netflix/titus/client/clustermembership/resolver/MultiNodeClusterMemberResolver.java

```diff
@@ -0,0 +1,286 @@
+            return ClusterMembershipSnapshot.empty();
+        }
+        // Each cluster member should be reported by by each resolver.
+        return buildSnapshot(findHealthySnapshots());
+    }
+
+    @Override
+    public Flux<ClusterMembershipSnapshot> resolve() {
+        return snapshotEventObservable;
+    }
+
+    private List<DirectClusterMemberResolver> findHealthyMembers() {
+        return memberResolversByIpAddress.values().stream().filter(DirectClusterMemberResolver::isHealthy).collect(Collectors.toList());
+    }
+
+    private List<ClusterMembershipSnapshot> findHealthySnapshots() {
+        return findHealthyMembers().stream().map(DirectClusterMemberResolver::getSnapshot).collect(Collectors.toList());
+    }
+
+    private ClusterMembershipSnapshot buildSnapshot(List<ClusterMembershipSnapshot> healthySnapshots) {
+        ClusterMembershipSnapshot.Builder builder = ClusterMembershipSnapshot.newBuilder();
+
+        Map<String, List<ClusterMembershipRevision<ClusterMember>>> grouped = healthySnapshots.stream()
+                .flatMap(snapshot -> snapshot.getMemberRevisions().values().stream())
+                .collect(Collectors.groupingBy(m -> m.getCurrent().getMemberId()));
+
+        List<ClusterMembershipRevision<ClusterMember>> recentRevisions = grouped.values().stream()
+                .map(this::findBestMemberRevision)
+                .collect(Collectors.toList());
+        builder.withMemberRevisions(recentRevisions);
+
+        // Find leader
+        Optional<ClusterMembershipRevision<ClusterMemberLeadership>> recentLeader = Optional.empty();
+        for (ClusterMembershipSnapshot snapshot : healthySnapshots) {
+            if (snapshot.getLeaderRevision().isPresent()) {
+                if (recentLeader.isPresent()) {
+                    if (recentLeader.get().getRevision() < snapshot.getLeaderRevision().get().getRevision()) {
+                        recentLeader = snapshot.getLeaderRevision();
+                    }
+                } else {
+                    recentLeader = snapshot.getLeaderRevision();
+                }
+            }
+        }
+        recentLeader.ifPresent(builder::withLeaderRevision);
+
+        // Choose latest version of each
+        long minStaleness = healthySnapshots.stream()
+                .mapToLong(ClusterMembershipSnapshot::getStalenessMs)
+                .min()
+                .orElse(0);
+        builder.withStalenessMs(minStaleness);
+
+        return builder.build();
+    }
+
+    private ClusterMembershipRevision<ClusterMember> findBestMemberRevision(List<ClusterMembershipRevision<ClusterMember>> singleMemberVersions) {
+        Preconditions.checkArgument(singleMemberVersions.size() > 0);
+
+        long bestRevision = singleMemberVersions.get(0).getRevision();
+        ClusterMembershipRevision<ClusterMember> best = singleMemberVersions.get(0);
+
+        for (int i = 1; i < singleMemberVersions.size(); i++) {
```

Comment:
minor nit: `List` does not guarantee random access is O(1), changing this code to using an iterator would be safer